libyul/Object.cpp

```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
 	return qualifiedNames;
 }
 
-std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+void Object::visitPath(YulString _qualifiedName, std::function<bool(Object const*)> const& _visitor) const
 {
-	yulAssert(_qualifiedName != name, "");
-	yulAssert(subIndexByName.count(name) == 0, "");
-
 	if (boost::algorithm::starts_with(_qualifiedName.str(), name.str() + "."))
 		_qualifiedName = YulString{_qualifiedName.str().substr(name.str().length() + 1)};
 	yulAssert(!_qualifiedName.empty(), "");
 
 	std::vector<std::string> subObjectPathComponents;
 	boost::algorithm::split(subObjectPathComponents, _qualifiedName.str(), boost::is_any_of("."));
 
-	std::vector<size_t> path;
 	Object const* object = this;
 	for (std::string const& currentSubObjectName: subObjectPathComponents)
 	{
 		yulAssert(!currentSubObjectName.empty(), "");
-		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
-		yulAssert(
-			subIndexIt != object->subIndexByName.end(),
-			"Assembly object <" + _qualifiedName.str() + "> not found or does not contain code."
-		);
-		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
-		yulAssert(object, "Assembly object <" + _qualifiedName.str() + "> not found or does not contain code.");
-		yulAssert(object->subId != std::numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+
+		YulString subObjectName = YulString{currentSubObjectName};
+		object = object->at(subObjectName);
+		if (object && _visitor(object))
+			break;
 	}
+}
 
+std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+{
+	std::vector<size_t> path;
+	yulAssert(_qualifiedName != name, "");
+	yulAssert(subIndexByName.count(name) == 0, "");
+
+	this->visitPath(_qualifiedName, [&](Object const* _object) -> bool {
+		yulAssert(_object->subId != std::numeric_limits<size_t>::max(), "");
+		path.push_back(_object->subId);
+		return false;
+	});
 	return path;
 }
+
+Object const* Object::subObjectAt(YulString _qualifiedName)
+{
+	yulAssert(!_qualifiedName.empty(), "");
+
+	// If there is no `.` in the given `_qualifiedName`, the target
+	// object name is considered to be equal to the `_qualifiedName`, otherwise,
+	// the target object name is the last element in the path given by `_qualifiedName`.
+	YulString targetObjectName = _qualifiedName;
+	size_t targetObjectPos = _qualifiedName.str().find_last_of(".");
+	if (targetObjectPos != std::string::npos)
+		targetObjectName = YulString(_qualifiedName.str().substr(targetObjectPos + 1));
+
+	Object const* foundObject = nullptr;
+	this->visitPath(_qualifiedName, [&](Object const* _subObject) -> bool {
+		if (targetObjectName != _subObject->name)
+			return false;
+		foundObject = _subObject;
+		return true;
+	});
+
+	return foundObject;
```

Comment:
Note that `_qualifiedName` is not modified, and this function expects it to be always in the format: `(A.*)+`, where `A` is a Yul object name. The search is performed using `_qualifiedName`, and `targetObjectName` is the name of the sub-object we are looking for. It will be `nullptr` only if the object is not found, i.e., it is not a sub-object of the top-level object.

Hum... I don't think we always expect the `_qualifiedName` to produce valid sub object, since this is an user input. Some asserts are already in place to validate parts of it, but without test coverage, I'm unsure if we are covering everything. For instance, these inputs are invalid and fail as expected:

```
./build/test/tools/yulopti --object "A." test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object ".A" test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "." test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "..." test/libyul/objectCompiler/subObjectAccess.yul
# A is not a sub object of B
./build/test/tools/yulopti --object "B.A" test/libyul/objectCompiler/subObjectAccess.yul
```

Reply:
Should I validate it with a regex?

Reply:
> Should I validate it with a regex? Although it could be any valid Yul Object identifier, so not really sure if we need more validation than we already have.

Nope, no need.

Reply:
Having asserts instead of validations here is fine for now - in the compiler we have those validations already elsewhere and yulopti is an internal tool. I mean, I'd include a validation if I were writing it, but not having one is not a big deal as long as it has a nice message.

Just, if we're using asserts, I'd try to make it somewhat consistent with how other functions behave. Now `subObjectAt()`, `visitPath()`, and `pathToSubObject()` each has a different idea of what to do on objects that do not exist and on `Data` objects. Since `pathToSubObject()` already existed and we use it elsewhere in the compiler, IMO it would be sensible if all of them followed its behavior (i.e. just assert against all of these cases).